src/components/explorer/suggestion/SuggestionCombobox.tsx

```diff
@@ -14,76 +13,226 @@ import {
-    e.stopPropagation();
+  const changeTab = (e: ChangeEvent<HTMLInputElement>, tab: FocusTab) => {
+    if (!e.target.checked) return;
+    setFocusTab(tab as FocusTab);
+    const el = document.getElementById(`filtered-${tab}-list`);
+    const y = window.scrollY;
+    const x = window.scrollX;
+    el?.scrollIntoView({ behavior: 'smooth', block: 'start' });
+    // scrollIntoView move all scroll container, we need to reset main scroll to prevent double scrolling
+    window.scroll(x, y);
+  };
 
-    if (e.key === 'Home') selectFirstOption(root.current);
-    if (e.key === 'End') selectLastOption(root.current);
-    if (e.key === 'ArrowDown') selectNextSibling(root.current);
-    if (e.key === 'ArrowUp') selectPreviousSibling(root.current);
+  const onInput = (e: ChangeEvent<HTMLInputElement>) => {
+    setOpen(true);
+    setSearch(e.currentTarget.value);
   };
 
   const suggestionListProps = {
     setOpen,
     listboxId,
     filteredPairs,
+    filteredTokens,
   };
 
+  const filters = {
+    token: filteredTokens,
+    pair: filteredPairs,
+  };
+
+  useEffect(() => {
+    if (!open) return;
+    const observer = new IntersectionObserver((entries) => {
+      for (const entry of entries) {
+        if (entry.isIntersecting && entry.intersectionRatio > 0) {
+          const tab = (entry.target as HTMLElement).dataset.tab;
+          if (tab) setFocusTab(tab as FocusTab);
+        }
+      }
+    });
+    for (const tab in tabs) {
+      const el = document.getElementById(`filtered-${tab}-list`);
+      if (el) observer.observe(el);
+    }
+    return () => observer.disconnect();
+  }, [open, filteredTokens.length, filteredPairs.length]);
+
+  useEffect(() => {
+    if (!open) return;
+    const listbox = document.getElementById(listboxId);
+    const keydownHandler = (e: KeyboardEvent) => {
+      if (open && e.key === 'Escape') {
+        (e.target as HTMLInputElement).value = '';
+        return setOpen(false);
+      }
+      if (!open) return setOpen(true);
+
+      if (e.key === 'Escape') return setOpen(false);
+      if (e.key === 'Enter') return selectCurrentOption(listbox);
+
+      if (!['ArrowDown', 'ArrowUp', 'Home', 'End'].includes(e.key)) return;
+      e.preventDefault();
+      e.stopPropagation();
+
+      if (e.key === 'Home') selectFirstOption(listbox);
+      if (e.key === 'End') selectLastOption(listbox);
+      if (e.key === 'ArrowDown') selectNextSibling(listbox);
+      if (e.key === 'ArrowUp') selectPreviousSibling(listbox);
+    };
+    const handleSoftExit = (e: MouseEvent) => {
+      if (!root.current || !(e.target instanceof Element)) return;
+      if (!root.current.contains(e.target)) setOpen(false);
+    };
+    document.addEventListener('keydown', keydownHandler);
+    document.addEventListener('mousedown', handleSoftExit);
+    return () => {
+      document.removeEventListener('mousedown', handleSoftExit);
+      document.removeEventListener('keydown', keydownHandler);
+    };
+  });
```

Comment:
Just checking if you want this useEffect to run on every re-render or the list of dependencies is missing